spec/models/course/condition/achievement_spec.rb

```diff
@@ -20,11 +22,29 @@
 
       context "when an achievement is already included in its conditional's conditions" do
         subject do
-          create(:course_condition_achievement, :duplicate_child, course: course).tap do |cca|
+          create(:course_condition_achievement, course: course).tap do |cca|
             allow(cca).to receive(:achievement_id_changed?).and_return(true)
           end
         end
-        it { is_expected.to_not be_valid }
+
+        it 'is not valid' do
+          expect(subject).to_not be_valid
+          expect(subject.errors[:achievement]).to_not be_blank
+        end
+      end
+
+      context 'when an achievement is required by another conditional with the same id' do
```

Comment:
@minhtule I stared at this for a while, I finally get what you're trying to do.

Achievement conditions store a reference to their conditional, and the conditional is a polymorphic relation. So you want to make sure that two different polymorphic types with the same ID can exist. Is that what you're testing?

If that's what you're testing, what you should do is make the conditional_type and conditional_id columns a unique index, then this scenario doesn't really need to be tested.